### ISSUE TITLE
The implementation of towered fields is difficult to relate to documented constructions

### ISSUE THREAD

**Issue by daira** [arkworks-rs/algebra#264]
(This issue spans `algebra` and `curves`, but seems more pertinent here.)

Consider the implementation of towers of field extensions, for example [here](https://github.com/arkworks-rs/curves/blob/master/bls12_381/src/fields) for BLS12-381. The usual way to describe such towers is by the irreducible polynomial used for each extension. For instance, the [bls12_381](https://github.com/zkcrypto/bls12_381) crate implements F<sub>p<sup>2</sup></sub> as F<sub>p</sub>[u]/(u<sup>2</sup> + 1), and then F<sub>p<sup>6</sup></sub> as F<sub>p<sup>2</sup></sub>[v]/(v<sup>3</sup> - (u + 1)), and then F<sub>p<sup>12</sup></sub> as F<sub>p<sup>6</sup></sub>[w]/(w<sup>2</sup> - v).

(Strictly speaking there's an abuse of notation here because, e.g. when we write F<sub>p<sup>6</sup></sub>[w]/(w<sup>2</sup> - v) we're conflating the field F<sub>p<sup>6</sup></sub> with its representation, but apparently everyone abuses notation that way.)

The choice of irreducible polynomials is fairly arbitrary, but for interoperability it's best for the originator of a curve to pick them and for everyone to use the same ones for that curve. A specific representation for each extension field used by a pairing is needed in order to unambiguously specify the curve and generator for G<sub>2</sub>, and for test vectors. (You can technically convert between representations, but it's a pain and no-one should need to do that extra work.)

The problem is that the implementation of a tower in arkworks doesn't directly specify these irreducible polynomials (and also doesn't document what they are for existing implementations). Each extension specifies a residue and a non-residue, but I can't tell how these determine the polynomial, even after poring over the implementations in `algebra/ff/src/fields/models`.

**Comment by weikengchen** [comment]
I agree that the irreducible polynomial is basically hidden throughout both repos. And it seems that the current implementations in https://github.com/arkworks-rs/algebra/tree/master/ff/src/fields/models focus on only specific irreducible polynomials. 

Should we add more explanations in `models`? Do you think there will be use cases that use more special irreducible polynomials?

**Comment by weikengchen** [comment]
(Note: generally there exist some prime numbers that could cause some of these polynomials to be not irreducible, but it seems to be rare)

**Comment by daira** [comment]
If I had to guess by extrapolation from BLS12-381, I'd guess that a k-extension uses the irreducible polynomial u<sup>k</sup> - `RESIDUE`. Is that right?

More explanation would definitely be good!

It's not so rare; e.g. Pluto's field cannot use u<sup>2</sup> + 1 for F<sub>p<sup>2</sup></sub>, and so it is specified to use u<sup>2</sup> + 5.

**Comment by daira** [comment]
This is blocking my implementation of Pluto and Triton in arkworks-rs/curves#54. The documentation needed is of how the constants such as `NONRESIDUE` and `QUADRATIC_NONRESIDUE` correspond to the usual description of a tower.

**Comment by Pratyush** [comment]
`QUADRATIC_NONRESIDUE` is used only for generating the square-root algorithm (I believe), while the `NONRESIDUE` is used to construct the extension.

**Comment by weikengchen** [comment]
So it seems that it is possible to implement u^2+5, just one needs to choose the NONRESIDUE in Fq2Parameters to be -5.

On a related note, I recently also implemented the BN446 inside the arkworks environment here: https://github.com/oblivious-file-sharing/netherite-algebra/tree/main/src/curve_bn446 
It has associated Sage scripts, which may be helpful.
